gh-87414: add musl support to platform.libc_ver

cmaureir opened the pull request:
This includes an implementation from the packaging module to detect the `musl` version on the system,
allowing `libc_ver` to properly return the version rather than an empty tuple.


<!-- gh-issue-number: gh-87414 -->
* Issue: gh-87414
<!-- /gh-issue-number -->

cmaureir committed:
bpo-43248: add musl support to platform.libc_ver

This includes an implementation from the packaging
module to detect the musl version on the system,
allowing libc_ver to properly return the version
rather than an empty tuple.

Co-authored-by: Manuel Kaufmann <humitos@gmail.com>

cmaureir committed:
Remove typehints and the MuslVersion structure

ambv committed:
Merge branch 'main' into musl_libc_ver

humitos committed:
The relative import is not valid

We get the following error otherwise:

```
ImportError: attempted relative import with no known parent package
```

humitos committed:
Remove typing annotations from `_elffile.py`

humitos committed:
Remove unused classes from `_elffile.py`

humitos committed:
Move ELFFile class from its own module to `platform.py`

humitos committed:
Add some tests for ELFFile class and musl functions

humitos committed:
Add `@support.requires_musl` decorator to skip tests if needed

Put the musl tests behind `requires_musl` decorator so they are not run in
systems without its support.

malemburg commented on Lib/platform.py:
```diff
@@ -199,6 +200,10 @@ def libc_ver(executable=None, lib='', version='', chunksize=16384):
         binary = f.read(chunksize)
         pos = 0
         while pos < len(binary):
+            if b'musl' in binary:
```
Please move this case into the if-else right below as elif after the more common libc / GLIBC case.

cmaureir replied:
The motivation to put it before was to check it as soon as possible to break the loop, but also because the `libc` bytes will be found in a  musl binary, so having:
```
if b'libc' in binary or b'GLIBC' in binary:
    ...
elif b'musl' in binary:
    ...
```
first will be true, and m will have a no match regular expression.

I think we either put first the musl condition, or keep it as it's at the moment.

What do you think @malemburg  ?

malemburg replied:
Fair enough. Please add a comment explaining this, though.

malemburg commented on Lib/platform.py:
```diff
@@ -1344,6 +1349,121 @@ def freedesktop_os_release():
+# These functions were copied and adapted from the packaging module:
+# https://github.com/pypa/packaging/blob/main/src/packaging/_musllinux.py
+# https://github.com/pypa/packaging/blob/main/src/packaging/_elffile.py
+
```
Please add a comment which clarified the license terms on this copied part. It seems that the packaging code is dual licensed under the Apache and BSD license, but it's not clear whether the logic actually originated in the package repo or whether it was copied from elsewhere.

cmaureir replied:
Looking at the reference that the code states:
https://gist.github.com/lyssdod/f51579ae8d93c8657a5564aefc2ffbca
it seems it was only an inspiration and not a 1:1 copy of the code.

I see the code was written by @uranusjr so hopefully we could get a clearer answer :)

uranusjr replied:
The code (for musl) was pretty much rewritten at some point and doesn’t really resemble the original inspiration anymore. If I remember correctly I was mostly reading the ELF documentation ([man page](https://man7.org/linux/man-pages/man5/elf.5.html) and [this page](https://refspecs.linuxfoundation.org/elf/gabi4+/ch4.eheader.html)) when I was doing it. The code in `_manylinux` resembles the `guess_runtime` code more previously, but that was mostly taken out after the rewrite.

malemburg replied:
Thanks for the added context. I think we're fine license-wise, then.

malemburg reviewed (changes requested):
Still a few minor things to fix, but overall, this looks good.

Please also add a NEWS entry.

Thanks.

humitos committed:
Add `NEWS.d` entry

malemburg commented on Lib/test/support/__init__.py:
```diff
@@ -387,6 +387,19 @@ def wrapper(*args, **kw):
+def requires_musl():
+    """Decorator raising SkipTest if the musl is not available."""
+    import subprocess
+    proc = subprocess.run(["ldd"], stderr=subprocess.PIPE, universal_newlines=True)
```
See my comment in the conversation section. The above fails on Macs since they don't ship with `ldd `as a tool.

mara004 commented on Lib/platform.py:
```diff
@@ -1344,6 +1349,121 @@ def freedesktop_os_release():
+### musl libc version support
+# These functions were copied and adapted from the packaging module:
+# https://github.com/pypa/packaging/blob/main/src/packaging/_musllinux.py
+# https://github.com/pypa/packaging/blob/main/src/packaging/_elffile.py
```
nit: If the packaging module switches to use this new stdlib code at some point, the references might become invalid. A possible solution would be to use permalinks.

cmaureir committed:
Merge branch 'main' into musl_libc_ver

cmaureir committed:
Adapt latest comments from review

- Making requires_musl compatible with macOS
- Add permalinks on the packaging references
- Add comment for the b'musl' check before b'libc'

encukou commented on Lib/test/test_platform.py:
```diff
@@ -538,6 +542,49 @@ def test_parse_os_release(self):
+    @support.requires_subprocess()
+    @support.requires_musl()
+    def test_libc_ver_musl(self):
+        self.assertEqual(platform.libc_ver(), ("musl", "1.2"))
```
Will this always be 1.2?
